### PR TITLE
Add test failure cause mapping

### LINKED ISSUE #26299: Gradle doesn't expose a comparison failure in test exception causes
### Current Behavior

Gradle submits a finish test event with an unrecognized nested Junit comparison failure (in the cause of the regular exception). This comparison exception is exposed as DefaultFailure.
image

### Expected Behavior

Gradle should submit a finish test event that has the structure:
```
DefaultTestFinishEvent {
    result = DefaultTestFailureResult {
        failures = [DefaultTestFailure {
            className = "AssertionError"
            causes = [DefaultTestAssertionFailure {
                className = "AssertionFailedError"
                expectedText = "Expected text"
                actualText = "Actual text"
            }]
        }]
    }
}
```

### Context (optional)

Intellij Idea used `ProgressListener` for receiving Gradle events. In particular, we receive test events with comparison info. This info we use for showing detailed comparison view with highlighting of same and different lines. Our users have tests that can wrap exceptions from assertEquals by other exceptions with some additional information about a test or a test environment.

### Steps to Reproduce

1. Create a project with the test file:
```
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opentest4j.AssertionFailedError

class TestCase {

    @Test
    fun `test wrapped comparison exception`() {
        try {
            Assertions.assertEquals("Expected text", "Actual text", "Assertion message")
        } catch (ex: AssertionFailedError) {
            throw AssertionError("Additional message", ex)
        }
    }
}
```
2. Create a project with Gradle tooling and configure `PregressListener` for `OperationType.TEST`:
```
public class GradleProgressListener implements ProgressListener {
  @Override
  public void statusChanged(ProgressEvent event) {
    if (event instanceof TestFinishEvent testFinishEvent) {
      if (testFinishEvent.getResult() instanceof FailureResult failureResult) {
        System.out.println("-----------");
        System.out.println("Result class: " + failureResult.getClass());
        for (Failure failure: failureResult.getFailures()) {
          System.out.println("Failure class: " + failure.getClass());
          if (failure instanceof TestFailure testFailure) {
            System.out.println("Failure exception: " + testFailure.getClassName());
            for (Failure cause: testFailure.getCauses()) {
              System.out.println("Cause class: " + cause.getClass());
              if (cause instanceof TestFailure causeFailure) {
                System.out.println("Cause exception: " + causeFailure.getClassName());
              }
            }
          }
        }
      }
    }
  }
}
```
3. Execute test task by BuildLauncher

Expected logs:
```
Result class: class org.gradle.tooling.events.test.internal.DefaultTestFailureResult
Failure class: class org.gradle.tooling.internal.consumer.DefaultTestFrameworkFailure
Failure exception: java.lang.AssertionError
Cause class: class org.gradle.tooling.internal.consumer.DefaultTestAssertionFailure
Cause exception: java.lang. AssertionFailedError
```
Actual logs:
```
Result class: class org.gradle.tooling.events.test.internal.DefaultTestFailureResult
Failure class: class org.gradle.tooling.internal.consumer.DefaultTestAssertionFailure
Failure exception: java.lang.AssertionError
Cause class: class org.gradle.tooling.internal.consumer.DefaultFailure
```

### Gradle version

8.4-20230905234741+0000

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by hegyibalint**
Closes #26299

### CODE REVIEW

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.TestFailureSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.BuildException
+import org.gradle.tooling.TestAssertionFailure
+
+@ToolingApiVersion(">=8.5")
+@TargetGradleVersion(">=8.5")
+class TestFailureProgressEventCrossVersionTest extends TestFailureSpecification {
```
Comment (HackerMadCat) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy:27]: Could you create a test case with comparison failure for Junit 4 and Junit 5?
```
    def "Test failure contains mapped test comparison causes using JUnit 5"() {
        given:
        setupJUnit5()
        file('src/test/java/org/gradle/JUnitTest.java') << '''
            package org.gradle;
            import org.junit.jupiter.api.Test;
            public class JUnitTest {
                @Test
                void test() {
                    throw new AssertionError(
                        "This exception wraps an assertion error",
                        new AssertionFailedError("This is a wrapped assertion error", "Expected text", "Actual text")
                    );
                }
            }
        '''
        def collector = new TestFailureEventCollector()

        when:
        runTestTaskWithFailureCollection(collector)

        then:
        thrown(BuildException)
        collector.failures.size() == 1
        collector.failures[0] instanceof TestFailure

        TestFailure failure = collector.failures[0]
        failure.message == "This exception wraps an assertion error"

        failure.causes.size() == 1
        failure.causes[0] instanceof TestAssertionFailure
        
        TestAssertionFailure assertionFailure = failure.causes[0]
        assertionFailure.message == "This is a wrapped assertion error"
        assertionFailure.expected == "Expected text"
        assertionFailure.actual == "Actual text"
    }
```

**Review by HackerMadCat — COMMENTED** [review]
See comments in code:
> Could you create a test case with comparison failure for Junit 4 and Junit 5?
> Why are you skipping the runtime exception?
> Could you use `TestFailure` instead of `TestAssertionFailure`?

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        '''
+        def collector = new TestFailureEventCollector()
+
+        when:
+        runTestTaskWithFailureCollection(collector)
+
+        then:
+        thrown(BuildException)
+        collector.failures.size() == 1
+        collector.failures[0] instanceof TestAssertionFailure
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy:60]: `TestFailure` [extends](https://github.com/gradle/gradle/blob/b8bdbec5c239ecc0c47e4bbd4f52290fbd37c0ec/subprojects/tooling-api/src/main/java/org/gradle/tooling/TestAssertionFailure.java#L28) `TestFailure`.
In this case, I'm interested in the stricter type, as I'm explicitly doing an assertion failure.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy (hunk 1 above)
Comment (hegyibalint) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy:27]: Yeah, can do!
Also another case which I missed: is the functionality correct, when the assertion exception is wrapped in a runtime exception?

Comment (hegyibalint) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/TestFailureProgressEventCrossVersionTest.groovy:27]: Added two new test cases

**Comment by donat** [comment]
@bot-gradle test RFN

**Review by donat — APPROVED** [review]
@hegyibalint LGTM. Please check with @HackerMadCat if we added coverage for all expected scenarios.

**Comment by HackerMadCat** [comment]
@donat, @hegyibalint added tests for cases that I have.

ps. I cannot resolve my conversations. I haven't rights to it.

**Comment by hegyibalint** [comment]
> @donat, @hegyibalint added tests for cases that I have.
> 
> ps. I cannot resolve my conversations. I haven't rights to it.

No problems. Thanks for the feedback!

**Comment by hegyibalint** [comment]
@bot-gradle test and merge

**Comment by donat** [comment]
@bot-gradle test and merge

**Pull request merged**